src/ln/channelmanager.rs

```diff
@@ -822,29 +827,51 @@ impl ChannelManager {
 				let id_option = channel_state.as_ref().unwrap().short_to_id.get(&short_channel_id).cloned();
 				let forwarding_id = match id_option {
-					None => {
+					None => { // unknown_next_peer
 						return_err!("Don't have available channel for forwarding as requested.", 0x4000 | 10, &[0;0]);
 					},
 					Some(id) => id.clone(),
 				};
-				if let Some((err, code, chan_update)) = {
+				if let Some((err, code, chan_update)) = loop {
 					let chan = channel_state.as_mut().unwrap().by_id.get_mut(&forwarding_id).unwrap();
-					if !chan.is_live() {
-						Some(("Forwarding channel is not in a ready state.", 0x1000 | 7, self.get_channel_update(chan).unwrap()))
-					} else {
-						let fee = amt_to_forward.checked_mul(self.fee_proportional_millionths as u64).and_then(|prop_fee| { (prop_fee / 1000000).checked_add(chan.get_our_fee_base_msat(&*self.fee_estimator) as u64) });
-						if fee.is_none() || msg.amount_msat < fee.unwrap() || (msg.amount_msat - fee.unwrap()) < *amt_to_forward {
-							Some(("Prior hop has deviated from specified fees parameters or origin node has obsolete ones", 0x1000 | 12, self.get_channel_update(chan).unwrap()))
-						} else {
-							if (msg.cltv_expiry as u64) < (*outgoing_cltv_value) as u64 + CLTV_EXPIRY_DELTA as u64 {
-								Some(("Forwarding node has tampered with the intended HTLC values or origin node has an obsolete cltv_expiry_delta", 0x1000 | 13, self.get_channel_update(chan).unwrap()))
-							} else {
-								None
-							}
-						}
+
+					if !chan.is_live() { // temporary_channel_failure
+						break Some(("Forwarding channel is not in a ready state.", 0x1000 | 7, self.get_channel_update(chan).unwrap()));
+					}
+					if *amt_to_forward < chan.get_their_htlc_minimum_msat() { // amount_below_minimum
+						break Some(("HTLC amount was below the htlc_minimum_msat", 0x1000 | 11, self.get_channel_update(chan).unwrap()));
 					}
-				} {
-					return_err!(err, code, &chan_update.encode_with_len()[..]);
+					let fee = amt_to_forward.checked_mul(self.fee_proportional_millionths as u64).and_then(|prop_fee| { (prop_fee / 1000000).checked_add(chan.get_our_fee_base_msat(&*self.fee_estimator) as u64) });
+					if fee.is_none() || msg.amount_msat < fee.unwrap() || (msg.amount_msat - fee.unwrap()) < *amt_to_forward { // fee_insufficient
+						break Some(("Prior hop has deviated from specified fees parameters or origin node has obsolete ones", 0x1000 | 12, self.get_channel_update(chan).unwrap()));
+					}
+					if (msg.cltv_expiry as u64) < (*outgoing_cltv_value) as u64 + CLTV_EXPIRY_DELTA as u64 { // incorrect_cltv_expiry
+						break Some(("Forwarding node has tampered with the intended HTLC values or origin node has an obsolete cltv_expiry_delta", 0x1000 | 13, self.get_channel_update(chan).unwrap()));
+					}
+					let cur_height = self.latest_block_height.load(Ordering::Acquire) as u32 + 1;
+					if msg.cltv_expiry <= cur_height + CLTV_EXPIRY_DELTA as u32 { // expiry_too_soon
```

Comment:
Hmm, dont think we need the full CLTV_EXPIRY_DELTA here. The spec doesn't actually say what to use here (as far as I see) and appears to assume that meeting the min_final_cltv_expiry will suffice for every node on the path. I think all we really care about here is that we don't end up in a situation where we early-force-close a channel (ChannelMonitor::would_broadcast_at_height) because of a much-too-soon CLTV (implying, obviously, that no expiration is already past). This means:
* for the inbound HTLC (aka msg.cltv_expiry) that we wont learn the payment preimage within channelmonitor::CLTV_CLAIM_BUFFER. This implies that the outbound HTLC CLTV must be at least CLTV_CLAIM_BUFFER before the inbound HTLC (trivially true due to CLTV_EXPIRY_DELTA being greater, though maybe we could add a comment/static assert indicating that is a requirement) and that the peer is able to do the update_fulfill_htlc/commitment/revoke dance within CLTV_EXPIRY_DELTA - CLTV_CLAIM_BUFFER blocks (which should never be an issue, though, again, we should probably add a comment noting that CLTV_EXPIRY_DELTA should always be a block or three greater than CLTV_CLAIM_BUFFER).
* for the outbound HTLC (aka outgoing_cltv_value) that we receive a update_fail_htlc and do the commitment/revoke dance prior to the HTLC's expiration minus CLTV_CLAIM_BUFFER. Now that isnt actually possible to ensure right now, but I think we can change the check in ChannelMonitor to only force-close outbound if we're a block or three *after* the actual expiration as long as CLTV_EXPIRY_DELTA is sufficiently larger than CLTV_CLAIM_BUFFER (or, I think, at least 2x CLTV_CLAIM_BUFFER cause that will give us CLTV_EXPIRY_DELTA blocks to get it confirmed, which is the meaning of CLTV_EXPIRY_DELTA).

All that to say, if we tweak ChannelMonitor::would_broadcast_at_height and ChannelMonitor::block_connected we can probably just check that outgoing_htlc_value is at least the current height + a few blocks and dont worry about anything else assuming the next hop will either immediately fail it due to it being too close to their desired final_cltv_expiry or not being able to meet their own CLTV_EXPIRY_DELTA or they will fail it because it expires soon.

Reply:
> for the outbound HTLC (aka outgoing_cltv_value) that we receive a update_fail_htlc and do the commitment/revoke dance prior to the HTLC's expiration minus CLTV_CLAIM_BUFFER.

I don't clearly understand this sentence. 

Did you mean the outbound HTLC need to be force-closed if update_fail_htlc/commitment_dance is *not* received received prior to *corresponding incoming* HTLC's expiration minus CLTV_CLAIM_BUFFER?

Reply:
The second, kinda. We need to force-close the channel if the upstream node hasn't done the update_fail_htlc/commitment_signed dance at least in enough time for us to claim the corresponding incoming HTLC. Right now we force-close if they haven't update_fail danced at least by a few blocks *before* the HTLC expires, but that seems like overkill, instead I think we should change it to only force-close if they haven't update_fail danced at least a few blocks *after* the HTLC expires, as long as we do it at least CLTV_CLAIM_BUFFER blocks prior to the incoming HTLC expiring, which we can ensure by ensuring that CLTV_EXPIRY_DELTA is at least 2x as large as CLTV_CLAIM_BUFFER.